ts/nix.ts

```diff
@@ -174,6 +196,72 @@ export function toHumanReadable(nixExpr: NixExpression): string {
   return renderInterpolatedString(nixExpr.str, toHumanReadable);
 }
 
+function collectFlakeDeps(nixExpr: NixExpression): Record<string, FlakeDep> {
+  const collect = (arr: Array<NixExpression>) =>
+    arr.reduce(
+      (acc, nixExpr) => ({ ...acc, ...collectFlakeDeps(nixExpr) }),
+      {} as Record<string, FlakeDep>,
+    );
+  switch (nixExpr.type) {
+    case "raw":
+      return collect(getInterpolations(nixExpr.raw));
+    case "list":
+      return collect(nixExpr.elements);
+    case "attrSet":
+      return collect(Object.values(nixExpr.elements));
+    case "strLit":
+      return collect(getInterpolations(nixExpr.str));
+    case "flakeDep":
+      return { [nixExpr.name]: nixExpr.dep };
+  }
+  checkExhaustiveness(nixExpr);
+}
+
+/**
+ * Renders the nix expression as the `outputs` of a flake file with all flake
+ * deps provided as inputs.
+ *
+ * Example:
+ * ```typescript
+ * renderFlakeFile(nixAttrSet({
+ *   foo: nixFlakeDep("foo-repo", { url: "http://example.org/foo" }),
+ *   bar: nixFlakeDep("bar-repo", { url: "http://example.org/bar" }),
+ * }))
+ *
+ * // Returns:
+ * // {
+ * //   inputs.foo-repo.url = "http://example.org/foo";
+ * //   inputs.bar-repo.url = "http://example.org/bar";
+ * //   outputs = { self, foo-repo, bar-repo }: {
+ * //     "foo" = foo-repo;
+ * //     "bar" = bar-repo;
+ * //   };
+ * // }
+ * ```
+ */
+export function renderFlakeFile(nixExpr: NixExpression): string {
+  const flakeDeps = collectFlakeDeps(nixExpr);
+  const quoteStr = (s: string) => renderNixExpression(nixStrLit(s));
+  // Not using `renderNixExpression(nixAttrSet(...))` to render the flake file
+  // since the flake file is a subset of a valid Nix attribute set (e.g. quoted
+  // keys are not allowed)
+  return (
+    "{" +
+    Object.entries(flakeDeps)
+      .map(([name, dep]) =>
+        dep.flake !== false
+          ? `inputs.${name}.url = ${quoteStr(dep.url)};`
+          : `inputs.${name} = { url = ${quoteStr(dep.url)}; flake = false; };`,
+      )
+      .join("\n") +
+    "\noutputs = { " +
+    ["self", ...Object.keys(flakeDeps)].join(", ") +
+    " }:" +
+    renderNixExpression(nixExpr) +
+    ";\n}"
+  );
+}
```

Comment:
Could we add some tests for this? Probably in typescript, I think. Some simple test (like the one from the docstring here), but also some tests where we call `nixFlakeDep` in a deeper nested structure, like in a list or an attribute set or both.